File: 김세희/주차 요금 계산.js
const getMinute = (time) => {
  const [hour, min] = time.split(":").map(Number);
  return hour * 60 + min;
};

function solution(fees, records) {
  var answer = [];
  const [defTime, defFee, rangeTime, rangeFee] = fees;
  const parkingLot = {};
  const totalTime = {};
  const totalFee = {};
  const endTime = getMinute("23:59");

  for (let record of records) {
    const [time, car, type] = record.split(" ");
    const totalMin = getMinute(time);
    if (type === "IN") {
      parkingLot[car] = [totalMin, true];
      if (!totalTime[car]) {
        totalTime[car] = 0;
      }
    } else {
      const start = parkingLot[car][0];
      const parkingTime = totalMin - start;
      totalTime[car] += parkingTime;
      parkingLot[car][1] = false;
    }
  }

  const restParking = Object.entries(parkingLot).filter(
    ([car, info]) => info[1]
  );

  for (const [car, info] of restParking) {
    const start = info[0];
    const parkingTime = endTime - start;
    totalTime[car] += parkingTime;
  }

  for (const [car, parkingTime] of Object.entries(totalTime)) {
    if (parkingTime <= defTime) {
      totalFee[car] = defFee;
    } else {
      totalFee[car] =
        defFee + Math.ceil((parkingTime - defTime) / rangeTime) * rangeFee;
    }
  }

  return Object.entries(totalFee)
    .sort((a, b) => +a[0] - b[0])
    .map(([car, fee]) => fee);
}

File: 이현홍/주차 요금 계산.py
def solution(fees, records):
    answer = []
    L = len(records)
    for idx in range(L):
        records[idx] = records[idx].split()
    records.sort(key=lambda x:x[1])
    time_dic = {key:0 for key in sorted(set(map(lambda x:x[1], records)))}
    
    idx = 0
    while idx < L:
        number = records[idx][1]
        h, m = map(int, records[idx][0].split(":"))
        start = h*60 + m
        end = 24*60 -1
        if idx == L-1 or records[idx+1][1] != records[idx][1]:
            idx += 1
        else:
            h, m = map(int, records[idx+1][0].split(":"))
            end = h*60 + m
            idx += 2
        time_dic[number] += end-start
    for key in sorted(time_dic.keys()):
        pay = fees[1]
        residue = max(0, time_dic[key] - fees[0])
        pay += residue // fees[2] * fees[3]
        if residue%fees[2]: pay += fees[3]
        answer.append(pay)
    return answer
